### ISSUE TITLE
fx test project

### ISSUE THREAD

**Issue by diqp2001** [diqp2001/base_infrastructure#85]
@claude

into this project
https://github.com/diqp2001/base_infrastructure/tree/main/src/application/managers/project_managers/test_project_backtest%20fx

you will adapt the code to do similar to this will keeping similar  structure rules used in the 

https://github.com/diqp2001/base_infrastructure/tree/main/src/application/managers/project_managers/test_project_backtest

please a claude.md file in the folder to explain the project structure 

and most important use the data in the following csv file :
https://github.com/diqp2001/base_infrastructure/tree/main/data/fx_data

Nice — below is a complete algorithm class you can drop into your existing misbuffet / QCAlgorithm-style framework. It:

Uses LightGBM (lightgbm.LGBMClassifier) to predict next-period direction for each currency (trained on returns + technical features).

Computes a mean-reversion z-score (distance from rolling mean) per currency.

Combines the ML prediction and mean-reversion signal into a single score.

Selects top 2 currencies to long and bottom 2 to short each rebalance step.

Trades via FX pairs (see mapping and comments) and sizes positions as equal-weight percentages of portfolio.

Retrains weekly (configurable) and handles missing data robustly.

Notes / assumptions (important):

I interpreted your currency list ['EUR','GBP','AUD','USD','MXN','JPY','CAD'] as currencies and trade them against USD. For currencies where USD is quote (EUR, GBP, AUD), we use EURUSD, GBPUSD, AUDUSD; for JPY/CAD/MXN we use USDJPY, USDCAD, USDMXN and invert position signs as needed so "long JPY" means being long the JPY currency.

If your backtest environment has different symbol names (e.g. EUR/USD, EURUSD_FX), adapt the currency_pair_for mapping accordingly.

This code uses LightGBM — ensure lightgbm is installed in your environment.

Here’s the class:

import logging
from datetime import timedelta, datetime
from typing import Dict, List

import numpy as np
import pandas as pd
import lightgbm as lgb

from application.services.misbuffet.common import QCAlgorithm, Resolution

logger = logging.getLogger(__name__)

class FX_LGBM_MeanReversion_Algorithm(QCAlgorithm):
    def initialize(self):
        super().initialize()

        # --- User-configurable settings ---
        self.currencies = ["EUR", "GBP", "AUD", "USD", "MXN", "JPY", "CAD"]
        # Map currency -> tradable FX symbol in your framework
        # NOTE: adjust symbol strings to match your backtester (e.g. 'EURUSD' or 'EUR/USD')
        self.currency_pair_for = {
            "EUR": "EURUSD",
            "GBP": "GBPUSD",
            "AUD": "AUDUSD",
            # USD is special (we'll treat USD direction via inversions across pairs)
            "USD": None,      # handled implicitly by other pairs
            "MXN": "USDMXN",
            "JPY": "USDJPY",
            "CAD": "USDCAD",
        }

        # Only include currencies that have a mapped pair
        self.tradable_currencies = [c for c in self.currencies if self.currency_pair_for.get(c)]
        self.universe = [self.currency_pair_for[c] for c in self.tradable_currencies]

        # Feature / training settings
        self.lookback_window = 60      # rolling window for features & zscore
        self.train_window = 252        # training history (days)
        self.retrain_interval = timedelta(days=7)
        self.last_train_time = None

        # Model container
        self.models: Dict[str, lgb.LGBMClassifier] = {}

        # Positioning
        self.n_long = 2
        self.n_short = 2
        self.target_leverage_per_side = 0.4  # max fraction of portfolio for longs (and for shorts)
                                              # each long will get target_leverage_per_side / n_long

        # Add FX securities to the algorithm
        self.my_securities = {}
        for pair in self.universe:
            try:
                sec = self.add_forex(pair, Resolution.DAILY)
                self.my_securities[pair] = sec
                self.log(f"Added FX pair {pair}")
            except Exception as e:
                self.log(f"Failed to add {pair}: {e}")
                self.my_securities[pair] = None

        # Initial training (if data exists)
        self._train_models(self.time)

    # ---------------------------
    # Helpers: history & prices
    # ---------------------------
    def _get_history_df(self, symbol: str, lookback_days: int, end_time: datetime) -> pd.DataFrame:
        """Request historical daily bars and return a DataFrame with index=time and columns: open, high, low, close, volume (if available)."""
        try:
            hist = self.history([symbol], lookback_days, Resolution.DAILY, end_time=end_time)
        except Exception as e:
            self.log(f"history call failed for {symbol}: {e}")
            return pd.DataFrame()

        if isinstance(hist, dict):
            # some frameworks return dict of DataFrames keyed by symbol
            df = hist.get(symbol, pd.DataFrame())
            if isinstance(df, pd.DataFrame):
                return df.copy()
            return pd.DataFrame()
        elif isinstance(hist, pd.DataFrame):
            return hist.copy()
        else:
            return pd.DataFrame()

    def _current_price(self, symbol: str):
        # Prefer current slice if available
        try:
            # many frameworks support self.Securities[symbol].Price or self.Securities[symbol].close
            if symbol in self.securities:
                return float(self.securities[symbol].price)
        except Exception:
            pass

        # fallback to history last close
        df = self._get_history_df(symbol, 2, self.time)
        if df.empty:
            return None
        if "close" in df.columns:
            return float(df["close"].iloc[-1])
        # attempt to find typical price column
        for col in ["price", "last"]:
            if col in df.columns:
                return float(df[col].iloc[-1])
        return None

    # ---------------------------
    # Feature engineering
    # ---------------------------
    def _prepare_features_df(self, hist: pd.DataFrame) -> pd.DataFrame:
        """Given a historical DataFrame with 'close', compute features and next-day return label."""
        df = hist.copy()
        if "close" not in df.columns:
            return pd.DataFrame()
        df = df.sort_index()
        df["return"] = df["close"].pct_change()
        df["ma_short"] = df["close"].rolling(5).mean()
        df["ma_long"] = df["close"].rolling(20).mean()
        df["ma_gap"] = (df["close"] - df["ma_long"]) / df["ma_long"]  # relative deviation
        df["volatility"] = df["return"].rolling(self.lookback_window).std()
        df["rsi"] = self._rsi(df["close"], 14)
        # forward return (next day's return)
        df["return_fwd1"] = df["return"].shift(-1)
        df = df.dropna()
        return df

    def _rsi(self, series: pd.Series, period: int = 14) -> pd.Series:
        delta = series.diff()
        up = delta.clip(lower=0.0)
        down = -1 * delta.clip(upper=0.0)
        ma_up = up.rolling(period).mean()
        ma_down = down.rolling(period).mean()
        rs = ma_up / (ma_down + 1e-12)
        rsi = 100 - (100 / (1 + rs))
        return rsi

    # ---------------------------
    # Train per-currency model
    # ---------------------------
    def _train_model_for_currency(self, currency: str, current_time: datetime):
        pair = self.currency_pair_for.get(currency)
        if not pair:
            return None

        hist = self._get_history_df(pair, self.train_window + 5, current_time)
        df = self._prepare_features_df(hist)
        if df.empty:
            self.log(f"No training data for {pair}")
            return None

        # features & label
        X = df[["return", "ma_gap", "volatility", "rsi"]]
        # label: 1 if next-day return > 0 (we let ML predict direction)
        y = (df["return_fwd1"] > 0).astype(int)

        # LightGBM classifier
        model = lgb.LGBMClassifier(n_estimators=200, random_state=42)
        try:
            model.fit(X, y)
            self.log(f"Trained model for {currency} ({pair}) on {len(X)} samples")
            return model
        except Exception as e:
            self.log(f"Model train failed for {pair}: {e}")
            return None

    def _train_models(self, current_time: datetime):
        for c in self.tradable_currencies:
            model = self._train_model_for_currency(c, current_time)
            if model is not None:
                self.models[c] = model
        self.last_train_time = current_time
        self.log(f"Models retrained at {current_time.date()}")

    # ---------------------------
    # Mean reversion z-score
    # ---------------------------
    def _zscore(self, series: pd.Series, window: int):
        ma = series.rolling(window).mean()
        std = series.rolling(window).std()
        return (series - ma) / (std + 1e-12)

    # ---------------------------
    # OnData / decision logic
    # ---------------------------
    def on_data(self, data):
        # Retrain periodicly
        if self.last_train_time is None or (self.time - self.last_train_time) >= self.retrain_interval:
            self._train_models(self.time)

        if not self.models:
            return

        currency_scores = {}

        for cur in self.tradable_currencies:
            pair = self.currency_pair_for[cur]
            if pair not in self.my_securities or self.my_securities[pair] is None:
                continue

            # Prepare last lookback_window+1 history for features and zscore
            hist = self._get_history_df(pair, self.lookback_window + 5, self.time)
            if hist.empty or "close" not in hist.columns or len(hist) < (self.lookback_window + 2):
                continue
            df = self._prepare_features_df(hist)
            if df.empty:
                continue

            X_live = df[["return", "ma_gap", "volatility", "rsi"]].iloc[-1:].fillna(0.0)
            model = self.models.get(cur)
            if model is None:
                continue

            prob_pos = float(model.predict_proba(X_live)[0][1])  # probability of next-day positive return

            # mean-reversion zscore: positive z -> price above mean -> expect down (revert)
            z = self._zscore(hist["close"], self.lookback_window).iloc[-1]
            # reversion expectation for next-day return ~ -z (if z>0 expect negative)
            reversion_expectation = -float(z)

            # Combine signals:
            # - If prob_pos high AND reversion_expectation positive => strong long
            # - If prob_pos low AND reversion_expectation negative => strong short
            # Combine with a simple weighted sum (weights tunable)
            ml_w = 0.6
            mr_w = 0.4
            # Normalize prob_pos to [-1,1] where 0=neutral
            ml_signal = (prob_pos - 0.5) * 2.0
            combined_score = ml_w * ml_signal + mr_w * reversion_expectation

            currency_scores[cur] = {
                "pair": pair,
                "prob_pos": prob_pos,
                "z": z,
                "combined_score": combined_score
            }

        if not currency_scores:
            return

        # Rank currencies by combined score
        sorted_items = sorted(currency_scores.items(), key=lambda kv: kv[1]["combined_score"], reverse=True)

        longs = [c for c, _ in sorted_items[: self.n_long]]
        shorts = [c for c, _ in sorted_items[-self.n_short:]]

        self.log(f"Long candidates: {longs}, Short candidates: {shorts}")

        # Convert currency picks to pair-level orders, adjusting sign for pairs where USD is base
        # Position sizing: equal-weight across longs and equal-weight across shorts
        long_weight_each = self.target_leverage_per_side / max(1, len(longs))
        short_weight_each = self.target_leverage_per_side / max(1, len(shorts))

        # First clear positions not in new target set
        target_pairs = set(self.currency_pair_for[c] for c in longs + shorts if self.currency_pair_for.get(c))
        # Set holdings for each tradable pair
        for cur in self.tradable_currencies:
            pair = self.currency_pair_for[cur]
            if pair is None:
                continue
            target_pct = 0.0
            if cur in longs:
                # Long the currency:
                # If the pair is XUSD (currency as base) -> buy pair -> positive target_pct
                # If pair is USDX (USD base, e.g. USDJPY) -> to long foreign currency we need to SHORT the pair (inverse)
                if pair.endswith("USD"):
                    target_pct = long_weight_each
                elif pair.startswith("USD"):
                    target_pct = -long_weight_each  # inverse
            elif cur in shorts:
                # Short the currency:
                if pair.endswith("USD"):
                    target_pct = -short_weight_each
                elif pair.startswith("USD"):
                    target_pct = short_weight_each

            # Place the order (using percent target to keep it simple)
            try:
                if pair in self.securities:
                    self.set_holdings(pair, target_pct)
                    self.log(f"Set holdings for {pair}: target_pct={target_pct:.3f}")
                else:
                    self.log(f"Pair {pair} not in securities (skip)")
            except Exception as e:
                self.log(f"Order failed for {pair}: {e}")

    # ---------------------------
    # Utility: logging wrapper
    # ---------------------------
    def log(self, msg: str):
        logger.info(msg)

Quick tuning suggestions

Feature set: add momentum (e.g., 10-day return), ATR, or more lags for the classifier to improve performance.

Label: currently next-day return sign. For mean reversion you could label reversion specifically (i.e., sign opposite of recent abnormal deviation).

Retraining cadence: tune retrain_interval (weekly is common) and train_window (252) depending on your data frequency.

Position sizing: I use target_leverage_per_side to cap long and short exposure. Adjust to your desired risk/leverage.

If you want to trade pure currency pairs (not USD-quoted only), adjust currency_pair_for to actual symbols in your data.

If you want I can:

adapt the class to a different FX symbol naming convention (e.g. EUR/USD),

change labels to explicit mean-reversion labels,

add transaction cost modeling or slippage,

or convert to use probabilities from LightGBM as input to a Black-Litterman or optimizer (like in your stock version).
